### ISSUE TITLE
[GPU] Replace placeholder GPU forward pass with real CUDA-accelerated neural network inference

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#317]
## Problem Description

The `GpuInferenceEngine::forward_gpu` function in `crates/bitnet-inference/src/gpu.rs` currently returns a placeholder result instead of performing actual GPU-accelerated neural network inference. This critical stub prevents the GPU backend from delivering real model inference, effectively making GPU acceleration non-functional.

## Environment

**Affected Files:**
- `crates/bitnet-inference/src/gpu.rs` (lines 395-411)

**Current Implementation:**
```rust
fn forward_gpu(&self, input: &BitNetTensor, _step: usize) -> Result<BitNetTensor> {
    let compute_start = Instant::now();
    
    // This is a simplified synchronous version
    // In a full async implementation, we would use model.read().await
    
    // For now, create a placeholder result
    let result = BitNetTensor::zeros(&[1, 32000], candle_core::DType::F32, &self.backend.device)?;
    
    // Update compute time metrics
    {
        let mut metrics = self.metrics.lock().unwrap();
        metrics.compute_time_ms = compute_start.elapsed().as_millis() as f64;
    }
    
    Ok(result)
}
```

## Root Cause Analysis

1. **Placeholder Implementation**: The function creates zero tensors instead of executing model inference
2. **Missing Model Integration**: No interaction with the actual `BitNetModel::forward()` method
3. **Unused Parameters**: The `input` tensor and `_step` parameter are ignored
4. **Architecture Gap**: Missing bridge between GPU backend and model forward pass
5. **Performance Impact**: False metrics due to placeholder computation time

## Impact Assessment

**Severity**: High - Core GPU functionality non-functional
**Affected Components**: 
- GPU inference engine
- CUDA acceleration pipelines  
- Performance benchmarking (false metrics)
- Production GPU deployments

**User Impact**:
- GPU acceleration appears to work but provides no real acceleration
- False performance metrics mislead optimization efforts
- Production GPU inference serves placeholder outputs

### Primary Implementation

Replace the placeholder with real GPU-accelerated model forward pass:

```rust
fn forward_gpu(&self, input: &BitNetTensor, step: usize) -> Result<BitNetTensor> {
    let compute_start = Instant::now();
    
    // Access the model for inference
    let model = self.model.blocking_read();
    
    // Convert input to appropriate device tensor
    let device_input = input.to_device(&self.backend.device)?;
    
    // Prepare or update KV cache for the current step
    let mut cache = self.cache.lock().unwrap();
    cache.update_for_step(step)?;
    
    // Execute forward pass on GPU
    let output = if self.config.enable_mixed_precision && self.backend.supports_mixed_precision() {
        self.forward_mixed_precision(&model, &device_input, step)?
    } else {
        model.forward(&device_input, &mut *cache)?
    };
    
    // Ensure output is on correct device
    let result = output.to_device(&self.backend.device)?;
    
    // Update real compute time metrics
    {
        let mut metrics = self.metrics.lock().unwrap();
        metrics.compute_time_ms = compute_start.elapsed().as_millis() as f64;
    }
    
    Ok(result)
}
```

### Supporting Changes Required

1. **Update Mixed Precision Forward Pass** (`forward_mixed_precision`):
```rust
fn forward_mixed_precision(
    &self,
    model: &Box<dyn Model<Config = BitNetConfig>>,
    input: &BitNetTensor,
    step: usize,
) -> Result<BitNetTensor> {
    // Convert input to FP16/BF16 for mixed precision
    let mixed_input = if self.backend.supports_tensor_cores() {
        input.to_dtype(candle_core::DType::BF16)?
    } else {
        input.to_dtype(candle_core::DType::F16)?
    };
    
    // Execute forward pass with mixed precision
    let output = model.forward(&mixed_input)?;
    
    // Convert back to F32 for consistency
    output.to_dtype(candle_core::DType::F32)
}
```

2. **Enhance KV Cache Integration**:
   - Ensure cache is properly migrated to GPU device
   - Support dynamic cache growth for variable sequence lengths
   - Add cache prefetching for improved performance

3. **Device Compatibility Validation**:
   - Verify tensor device placement before forward pass
   - Add automatic device migration with error handling
   - Validate GPU memory availability before allocation

### Alternative Approaches

1. **Async-First Implementation**: Use `model.read().await` for fully async operation
2. **Streaming Implementation**: Support token-by-token streaming generation
3. **Batch Processing**: Implement true GPU batch parallelism

### Phase 1: Core Forward Pass (Priority: High)
- [ ] Replace placeholder with real model.forward() call
- [ ] Implement proper device tensor management
- [ ] Add KV cache integration for GPU inference
- [ ] Update metrics to reflect real computation time

### Phase 2: Mixed Precision Support (Priority: High)  
- [ ] Implement FP16/BF16 tensor conversions
- [ ] Add Tensor Core utilization for supported GPUs
- [ ] Optimize memory bandwidth with mixed precision kernels
- [ ] Add precision-aware error handling

### Phase 3: Performance Optimization (Priority: Medium)
- [ ] Implement GPU memory pooling for tensor allocation
- [ ] Add CUDA stream management for overlapped execution
- [ ] Optimize KV cache memory layout for GPU access patterns
- [ ] Add kernel fusion opportunities

### Phase 4: Production Hardening (Priority: Medium)
- [ ] Add comprehensive error handling for GPU operations
- [ ] Implement automatic fallback to CPU on GPU errors  
- [ ] Add memory usage monitoring and limits
- [ ] Create GPU-specific benchmarking suite

### Unit Tests
```rust
#[test]
fn test_gpu_forward_pass_real_inference() {
    let engine = setup_gpu_engine_with_test_model();
    let input = create_test_input_tensor();
    
    let result = engine.forward_gpu(&input, 0).unwrap();
    
    // Verify non-zero output (not placeholder)
    assert!(!result.is_zeros());
    assert_eq!(result.device(), &engine.backend.device);
}

#[test] 
fn test_mixed_precision_forward_consistency() {
    let engine = setup_gpu_engine_with_mixed_precision();
    let input = create_test_input_tensor();
    
    let fp32_result = engine.forward_gpu(&input, 0).unwrap();
    let mixed_result = engine.forward_mixed_precision(&model, &input, 0).unwrap();
    
    // Results should be close (within mixed precision tolerance)
    assert_tensors_close(&fp32_result, &mixed_result, 1e-3);
}
```

### Integration Tests
- Cross-validation against CPU inference engine
- Performance regression testing vs. current placeholder
- Memory usage validation under sustained load
- Multi-GPU device compatibility testing

### Validation Criteria
- [ ] GPU forward pass produces non-placeholder results
- [ ] Output tensors are mathematically equivalent to CPU inference (within tolerance)
- [ ] GPU memory utilization metrics are accurate
- [ ] Performance metrics reflect real computation time
- [ ] Mixed precision mode shows measurable speedup on compatible hardware

## Risk Assessment

**High Risk Items**:
- GPU memory management complexity may introduce leaks
- Device synchronization issues could cause deadlocks
- Mixed precision may introduce numerical accuracy issues

**Mitigation Strategies**:
- Implement comprehensive GPU memory tracking
- Add timeout mechanisms for GPU operations
- Create numerical accuracy validation suite
- Provide CPU fallback for all GPU operations

## Related Issues

This issue depends on and relates to several other GPU backend stub implementations:

- GPU Backend Device Management: `stub-gpu-backend-ensure-gpu-tensor.md`
- GPU Memory Manager: `stub-gpu-memory-manager-*.md` 
- GPU Quantization Kernels: `stub-gpu-quantizer-*.md`
- Mixed Precision Support: `stub-gpu-backend-supports-mixed-precision.md`
- Tensor Core Utilization: `stub-gpu-backend-supports-tensor-cores.md`

### Must Have (AC1): Real GPU Inference
- [ ] `forward_gpu()` executes actual model forward pass on GPU
- [ ] Input tensors are processed through neural network layers
- [ ] Output logits are mathematically valid (non-zero, proper distribution)
- [ ] GPU device placement is maintained throughout inference

### Must Have (AC2): Performance Metrics Accuracy  
- [ ] Compute time metrics reflect real GPU kernel execution time
- [ ] GPU utilization metrics accurately represent hardware usage
- [ ] Memory allocation tracking shows actual GPU memory consumption
- [ ] Benchmarks show measurable performance vs. placeholder

### Must Have (AC3): Mixed Precision Support
- [ ] FP16/BF16 mode executes when enabled and supported
- [ ] Tensor Core utilization on compatible GPUs (Volta+)
- [ ] Automatic precision fallback for unsupported hardware
- [ ] Numerical accuracy maintained within acceptable tolerances

### Should Have (AC4): Error Handling & Robustness
- [ ] Graceful degradation when GPU resources are exhausted
- [ ] Automatic CPU fallback on GPU computation failures
- [ ] Comprehensive error reporting for debugging
- [ ] Memory leak prevention with proper cleanup

### Nice to Have (AC5): Advanced Optimizations
- [ ] CUDA stream parallelism for improved throughput
- [ ] Kernel fusion opportunities for reduced memory bandwidth
- [ ] Dynamic batch size optimization based on available GPU memory
- [ ] Async inference pipeline with overlapped computation

---

**Priority**: P0 (Critical)
**Complexity**: High  
**Estimated Effort**: 2-3 weeks
**Required Skills**: CUDA, GPU architecture, neural network inference pipelines

**Comment by EffortlessSteven** [comment]
This is a GPU/CUDA-specific feature that requires a CUDA runtime environment. Closing as a platform-specific roadmap item. GPU work will be tracked separately once a CUDA runner is available.
